uqcsbot/starboard.py

```diff
@@ -40,20 +60,138 @@ def _rm_big_ratelimit(self, id: int):
     
-    def _query_sb_message(self, recv: int) -> Optional[int]:
-        """ Get the starboard message ID corresponding to the recieved message """
+    async def _blacklist_log(self, message: discord.Message, user: discord.Member, blacklist: bool):
+        channels = self.bot.get_all_channels()
+        modlog = discord.utils.get(channels, name="admin-alerts")
+        state = "blacklisted" if blacklist else "whitelisted"
+
+        await modlog.send(f"{str(user)} {state} message {message.id} (link: {message.jump_url})")
+    
+    @app_commands.default_permissions(manage_messages=True)
+    async def context_blacklist_sb_message(self, interaction: discord.Interaction, message: discord.Message):
+        """ Blacklists a message from being starboarded. If the message is already starboarded, also deletes it. """
+        db_session = self.bot.create_db_session()
+        # can't use the db query functions for this, they error out if a message hits the blacklist
+        entry = db_session.query(models.Starboard).filter(models.Starboard.recv == message.id)
+        query_val = entry.one_or_none()
+
+        if query_val is not None:
+            if query_val.sent is None:
+                # if the table has (recv, none) then it's already blacklisted.
+                return
+
+            # otherwise the table has (recv, something), we should delete the something and then make it (recv, none)
+            await (await self.starboard_channel.fetch_message(query_val.sent)).delete()
+            query_val.sent = None
+        else:
+            # other-otherwise the table doesn't have recv, so we add (recv, none)
+            db_session.add(models.Starboard(recv=message.id, recv_location=message.channel.id, sent=None))
+
+        db_session.commit()
+        db_session.close()
+
+        await self._blacklist_log(message, interaction.user, blacklist=True)
+        await interaction.response.send_message(f"Blacklisted message {message.id}.")
+
+    @app_commands.default_permissions(manage_messages=True)
+    async def context_whitelist_sb_message(self, interaction: discord.Interaction, message: discord.Message):
+        """ Removes a message from the starboard blacklist.
+            N.B. Doesn't perform an 'update' of the message. This may result in messages meeting the threshold
+            but not being starboarded if they don't get any more reacts. """
+        db_session = self.bot.create_db_session()
+
+        # if we find a (recv, none) for this message, delete it. otherwise the message is already not blacklisted.
+        entry = db_session.query(models.Starboard).filter(
+            models.Starboard.recv == message.id,
+            models.Starboard.sent == None
+        )
+        if entry.one_or_none() is not None:
+            entry.delete(synchronize_session=False)
+            db_session.commit()
+        db_session.close()
+
+        await self._blacklist_log(message.id, blacklist=False)
```

Comment:
This is missing interaction.user and needs to reference the message directly.

```suggestion
        await self._blacklist_log(message, interaction.user, blacklist=False)
```

Reply:
so true